In [2]:
import pandas as pd
import numpy as np

In [3]:
# read the dataset Nepal housing dataset
dataset = pd.read_csv("./Dataset/2020-4-27.csv")
dataset.head()

,Title,Address,City,Price,Bedroom,Bathroom,Floors,Parking,Face,Year,Views,Area,Road,Road Width,Road Type,Build Area,Posted,Amenities
0,Flat System House On Sale At Budhanikantha : H...,"Budhanikantha, Budhanilkantha, Kathmandu",Kathmandu,90000000,6,3,2.0,10,West,2073.0,17,1-0-0-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,18-0-0-0 Aana,11 hours ago,"['Parking', 'Lawn', 'Garage', 'Frontyard', 'Fe..."
1,21 Aana Beautiful House : House For Sale In Bu...,"Pasikot, Budhanilkantha, Kathmandu",Kathmandu,80000000,5,3,2.0,9,East,2073.0,26,0-21-0-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,0-14-0-0 Aana,12 hours ago,"['Parking', 'Garage', 'Fencing', 'Frontyard', ..."
2,Beautiful To Floor House : House For Sale In D...,"Dhapasi, Dhapasi, Kathmandu",Kathmandu,70000000,5,3,2.0,12,East,2071.0,154,1-1-0-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,10-1-0-0 Aana,2 days ago,"['Parking', 'Lawn', 'Garage', 'Frontyard', 'Fe..."
3,Modern Architecture Bungalow House : House For...,"baluwatar, Baluwatar, Kathmandu",Kathmandu,130000000,6,4,3.0,9,West,2017.0,17,0-10-1-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,0-7-1-0 Aana,1 day ago,"['Parking', 'Lawn', 'Garage', 'Frontyard', 'Fe..."
4,Modern Bungalow House : House For Sale In Golf...,"Golfutar, Golfutar, Kathmandu",Kathmandu,100000000,6,3,2.0,10,East,2070.0,59,0-19-2-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,0-14-2-0 Aana,2 days ago,"['Parking', 'Lawn', 'Garage', 'Balcony', 'Back..."


In [4]:
# check the data info
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2211 entries, 0 to 2210
Data columns (total 18 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Title       2211 non-null   object 
 1   Address     2211 non-null   object 
 2   City        2211 non-null   object 
 3   Price       2211 non-null   int64  
 4   Bedroom     2211 non-null   int64  
 5   Bathroom    2211 non-null   int64  
 6   Floors      1039 non-null   float64
 7   Parking     2211 non-null   int64  
 8   Face        2211 non-null   object 
 9   Year        582 non-null    float64
 10  Views       2211 non-null   object 
 11  Area        2211 non-null   object 
 12  Road        2211 non-null   object 
 13  Road Width  2211 non-null   object 
 14  Road Type   1426 non-null   object 
 15  Build Area  2211 non-null   object 
 16  Posted      2211 non-null   object 
 17  Amenities   2211 non-null   object 
dtypes: float64(2), int64(4), object(12)
memory usage: 311.0+ KB


`The dataset consists of 18 columns, with Price as the dependent variable and the remaining 17 as independent variables. The dataset contains a total of 2,211 records.`

In [11]:
# percentage of missing value 
dataset.isnull().sum()/dataset.shape[0]*100 

Title          0.000000
Address        0.000000
City           0.000000
Price          0.000000
Bedroom        0.000000
Bathroom       0.000000
Floors        53.007689
Parking        0.000000
Face           0.000000
Year          73.677069
Views          0.000000
Area           0.000000
Road           0.000000
Road Width     0.000000
Road Type     35.504297
Build Area     0.000000
Posted         0.000000
Amenities      0.000000
dtype: float64

In [13]:
for col in dataset.columns:
    if dataset[col].isnull().sum()>0:
        print(col)

Floors
Year
Road Type


In [17]:
# step 1
# Missing ra Non-missing grop bich ma Price ma difference xa ki xaina ?
missing_floors = dataset[dataset['Floors'].isnull()]['Price']
non_missing_floors = dataset[dataset['Floors'].notnull()]['Price']
print("missing_floors_mean:",missing_floors.mean())
print("non_missing_floors_mean: ", non_missing_floors.mean())

In [31]:
# step 2 
from scipy.stats import mannwhitneyu
stas,p = mannwhitneyu(missing_floors, non_missing_floors)
if p < 0.05:
    print( "Significant Difference -> missing important - add indicator, missing हुनु र Price बीच सम्बन्ध छ (missing important छ) ")
else :
    print("Not significant - missing random-Drop/fill")
    

Significant Difference -> missing important - add indicator, missing हुनु र Price बीच सम्बन्ध छ (missing important छ) 


##### floor Missing huda price High xa 
・ yo difference real ho ki chance le matra ayeko ho test garam

In [35]:
# Example DataFrame
from scipy.stats import ttest_ind
df = dataset.copy()
columns_to_check = ['Floors', 'Year', 'Road Type']  # Missing check गर्ने columns
target_col = 'Price'                        # Target column

# Result store गर्ने list
results = []

for col in columns_to_check:
    missing_group = df[df[col].isnull()][target_col]
    non_missing_group = df[df[col].notnull()][target_col]
    
    # Check if both groups have enough data
    if len(missing_group) > 1 and len(non_missing_group) > 1:
        t_stat, p_value = mannwhitneyu(missing_group, non_missing_group)
        missing_avg = missing_group.mean()
        non_missing_avg = non_missing_group.mean()
        diff = abs(missing_avg - non_missing_avg)
        
        results.append({
            'Column': col,
            'Missing Avg': missing_avg,
            'Non-Missing Avg': non_missing_avg,
            'Diff': diff,
            'T-stat': t_stat,
            'P-value': p_value
        })
    else:
        results.append({
            'Column': col,
            'Missing Avg': np.nan,
            'Non-Missing Avg': np.nan,
            'Diff': np.nan,
            'T-stat': np.nan,
            'P-value': np.nan
        })

# Results देखाउने
results_df = pd.DataFrame(results)
print(results_df)
    

      Column   Missing Avg  Non-Missing Avg          Diff    T-stat  \
0     Floors  6.491178e+11     1.327036e+11  5.164142e+11  333101.5   
1       Year  3.856829e+11     4.645491e+11  7.886623e+10  273273.5   
2  Road Type  9.938175e+10     5.754771e+11  4.760954e+11  457707.0   

        P-value  
0  1.179569e-75  
1  4.306651e-52  
2  1.240949e-12  


In [40]:
for col, p_value in zip(results_df['Column'], results_df['P-value']):
    if p_value < 0.05:
        print(f"{col} -> Significant Difference: missing important, add indicator")
    else:
        print(f"{col} -> Not Significant: drop or impute")

Floors -> Significant Difference: missing important, add indicator
Year -> Significant Difference: missing important, add indicator
Road Type -> Significant Difference: missing important, add indicator
